src/module_3/push_exploration.md

```diff
@@ -0,0 +1,5364 @@
+plt.plot([0, 1], [0, 1], color='navy', linestyle='--')
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('False Positive Rate')
+plt.ylabel('True Positive Rate')
+plt.title('ROC')
+plt.legend(loc="lower right")
+
+
+# Precision-Recall Curve
+precision_val, recall_val, _ = precision_recall_curve(y_val, y_val_probs)
+
+plt.subplot(1, 2, 2)
+plt.plot(recall_val, precision_val)
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('Recall')
+plt.ylabel('Precision')
+plt.title('Precision-Recall Curve')
+
+plt.tight_layout()
+plt.show()
+```
+
+
+    
+![png](push_exploration_files/push_exploration_48_0.png)
+    
+
+
+We obtain basically very similar results, so for the shake of simplicity, we will stay with the weight class that scikit provides. Curiously, both results obtain similar curves vs with no balance class for logistic regression. It may be due to the fact that, given the class 1 is so umbalanced, we do not appreciate high differences. However, we can see that in the precision in purchases (60% vs 4%, normal regression accuracy for positive is better) and recall in purchases (0% vs 62%, simple logistic regression does not find any purchase, while the later models do).
+
+Now, with these results, let's pause for a moment and analyze the numbers obtained:
+
+- Precision (True positive / true positive + false positive). Measures the accuracy of positive predictions. If the model predicts a purchase won't happen, it is correct almost always. If it predicts it will happen, only is correct 4% of the time (generating lots of false positives). Remember that current push open rate is around 5%, so it is feasible to be slightly below that. Would be useful to have the purchase rate per opened notification.
+
+- Recall or true positive rate (True positive / true positive + false negative). Measures the ability to identify all positives. If model predicts no purchase, it identifies 77% of the time. If the model predicts a purchase, it identifies 63% of the purchases
+
+- ROC Curve. Knowing that classes are strongly unbalanced, we can appreciate a few things: We start to generate false alarms very quickly, as provided by precision before. When 50% of the alarms are false (FPR at 0.5), we are identifying 80% of the positive alarms. Then the number of false alarms grow very quickly (so basically we will be spamming the user)
+
+- Precision-recall curve. Precision decreases very fast for almost no recall (loss of accuracy of positive predictions when trying to gain sensivity).
+
+We want to have a recall enough to appreciate purchases, and with a reasonable amount of false positives but fixed for avoiding churn rates (which Sales team mentioned it had a huge cost). Since we would like to target those users that are likely to buy those products, we can expect to increase the probability threshold and capture secure notifications. We will be missing positives, but sending those that we are more sure they will be buying.
```

Comment:
I am missing a truly business metric: business stakeholders in general don't know anything about recall or precision, or PR, just money